pkg/k8s/watchers/cilium_envoy_config.go

```diff
@@ -9,87 +9,189 @@ import (
+}
+
+func newCiliumEnvoyConfigWatcher(logger logrus.FieldLogger,
+	policyUpdater *policy.Updater,
+	serviceManager service.ServiceManager,
+	proxy *proxy.Proxy,
+	xdsServer envoy.XDSServer,
+	backendSyncer *envoy.EnvoyServiceBackendSyncer,
+) *ciliumEnvoyConfigWatcher {
+	return &ciliumEnvoyConfigWatcher{
+		logger:         logger,
+		policyUpdater:  policyUpdater,
+		serviceManager: serviceManager,
+		proxy:          proxy,
+		xdsServer:      xdsServer,
+		backendSyncer:  backendSyncer,
+		appliedCECs:    map[resource.Key]*ciliumv2.CiliumEnvoyConfig{},
+		appliedCCECs:   map[resource.Key]*ciliumv2.CiliumClusterwideEnvoyConfig{},
+	}
+}
+
+func (r *ciliumEnvoyConfigWatcher) handleCECEvent(ctx context.Context, event resource.Event[*ciliumv2.CiliumEnvoyConfig]) error {
+	scopedLogger := r.logger.
+		WithField(logfields.K8sNamespace, event.Key.Namespace).
+		WithField(logfields.CiliumEnvoyConfigName, event.Key.Name)
+
+	var err error
+
+	switch event.Kind {
+	case resource.Upsert:
+		scopedLogger.Debug("Received CiliumEnvoyConfig upsert event")
+		err = r.cecUpserted(ctx, event.Key, event.Object)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle CEC upsert")
+			err = fmt.Errorf("failed to handle CEC upsert: %w", err)
+		}
+	case resource.Delete:
+		scopedLogger.Debug("Received CiliumEnvoyConfig delete event")
+		err = r.cecDeleted(ctx, event.Key)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle CEC delete")
+			err = fmt.Errorf("failed to handle CEC delete: %w", err)
+		}
+	}
+
+	event.Done(err)
+
+	return nil
+}
+
+func (r *ciliumEnvoyConfigWatcher) handleCCECEvent(ctx context.Context, event resource.Event[*ciliumv2.CiliumClusterwideEnvoyConfig]) error {
+	scopedLogger := r.logger.
+		WithField(logfields.K8sNamespace, event.Key.Namespace).
+		WithField(logfields.CiliumClusterwideEnvoyConfigName, event.Key.Name)
+
+	var err error
+
+	switch event.Kind {
+	case resource.Upsert:
+		scopedLogger.Debug("Received CiliumClusterwideEnvoyConfig upsert event")
+		err = r.ccecUpserted(ctx, event.Key, event.Object)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle CCEC upsert")
+			err = fmt.Errorf("failed to handle CCEC upsert: %w", err)
+		}
+	case resource.Delete:
+		scopedLogger.Debug("Received CiliumClusterwideEnvoyConfig delete event")
+		err = r.ccecDeleted(ctx, event.Key)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle CCEC delete")
+			err = fmt.Errorf("failed to handle CCEC delete: %w", err)
+		}
+	}
+
+	event.Done(err)
+
+	return nil
+}
+
+func (r *ciliumEnvoyConfigWatcher) cecUpserted(ctx context.Context, key resource.Key, cec *ciliumv2.CiliumEnvoyConfig) error {
+	r.appliedCECsMutex.Lock()
+	defer r.appliedCECsMutex.Unlock()
+
+	appliedCEC, exists := r.appliedCECs[key]
+	if exists {
+		return r.updateCiliumEnvoyConfig(appliedCEC.ObjectMeta, appliedCEC.Spec, cec.ObjectMeta, cec.Spec)
```

Comment:
does this properly handle the case where have the sequence 
```
1. Add A
2. Add B
3. Update A (fails temporarily)
4. Update B
5. Update A (retry)
```

Asked differently, are the individual config objects independent from each other, or can they overwrite each other? Asking because the default behaviour of Resource is to retry on error, and envoy sounds like a blob of state that might not like that.

Reply:
Good point.

When parsing the Envoy resources from a `CiliumEnvoyConfig`, the resources are qualified by prefixing them with the namespace and name of their CEC. This way, different CEC's can't override the Envoy resources of each other. (See [ParseResources](https://github.com/cilium/cilium/blob/adeec1dfc6e2b25eabbf5a21f5cd76779a195cc1/pkg/envoy/ciliumenvoyconfig.go#L146) & [ResourceQualifiedName](https://github.com/cilium/cilium/blob/main/pkg/policy/api/utils.go#L149-L172) for furher info)

All "managed" cases within Cilium (where Cilium Operator is creating the `CiliumEnvoyConfig` (e.g. Ingress, Gateway API, L7LB)) work like this.

There are edge cases if a user is creating a `CiliumEnvoyConfig` by hand - **AND** uses slashes `/` in the Envoy resources name in the `CiliumEnvoyConfig`. This disables the previously mentioned prefixing process while parsing the resources. But this is already the case with the current implementation and can also happen in case of a successful reconciliation. This is a known challenge / trap when using the `CiliumEnvoyConfig`.

Reply:
sounds like you thought it through, that's good enough for me!